File: crates/frances-workflow/src/modules/workflow.rs
//! `frances:v1/workflow` — lifecycle (`exit`) + the busy-indicator
//! setter (`setStatus`).
//!
//! `exit()` requests graceful shutdown rather than killing the inbox
//! directly: it pulses `shutdown_notify`, and the `frances:v1/lifecycle`
//! module's background IIFE turns that into "run the user's shutdown
//! hook, then close the inbox." Workflows without a registered
//! `lifecycle.shutdown` handler still terminate promptly — the IIFE
//! closes the inbox unconditionally after the (absent) hook returns.
//!
//! `setStatus(text | null)` drives the TUI footer's busy indicator:
//! `Some(text)` shows the text with a spinner, `None` hides it.
//!
//! `setTitle(text | null)` sets or clears the session title. It rides
//! the same surfaces channel; the driver persists it into session
//! metadata (unlike the footer, it outlives the workflow).

use std::sync::Arc;

use rquickjs::function::Opt;
use rquickjs::{Ctx, Function, Result as JsResult};
use tokio::sync::Notify;
use tokio::sync::mpsc::UnboundedSender;

use crate::runtime::SurfaceCmd;

pub(crate) fn build_exit<'js>(
    ctx: &Ctx<'js>,
    shutdown_notify: Arc<Notify>,
) -> JsResult<Function<'js>> {
    Function::new(ctx.clone(), move || {
        shutdown_notify.notify_waiters();
        Ok::<_, rquickjs::Error>(())
    })
}

/// Build `setStatus(text | null)`. The argument is optional and
/// nullable: a string sets the indicator, `null`/`undefined`/omitted
/// clears it.
pub(crate) fn build_set_status<'js>(
    ctx: &Ctx<'js>,
    surfaces_tx: UnboundedSender<SurfaceCmd>,
) -> JsResult<Function<'js>> {
    Function::new(ctx.clone(), move |text: Opt<Option<String>>| {
        let cmd = match text.0.flatten() {
            Some(text) => SurfaceCmd::SetFooter { text },
            None => SurfaceCmd::ClearFooter,
        };
        let _ = surfaces_tx.send(cmd);
        Ok::<_, rquickjs::Error>(())
    })
}

/// Build the `_setTitle(text | null)` primitive. Same optional/nullable
/// argument shape as `setStatus`; the JS wrapper in
/// `assets/frances/v1/workflow.js` layers `getTitle()` on top by caching
/// the last value it sent.
pub(crate) fn build_set_title<'js>(
    ctx: &Ctx<'js>,
    surfaces_tx: UnboundedSender<SurfaceCmd>,
) -> JsResult<Function<'js>> {
    Function::new(ctx.clone(), move |text: Opt<Option<String>>| {
        let _ = surfaces_tx.send(SurfaceCmd::SetTitle {
            title: text.0.flatten(),
        });
        Ok::<_, rquickjs::Error>(())
    })
}

File: crates/frances-workflow/assets/frances/v1/workflow.js
// `frances:v1/workflow` — re-exports `exit` + `setStatus` from the
// install-time stash and layers `setTitle`/`getTitle` over the host's
// `_setTitle` primitive. The stash is deleted from globalThis after all
// modules evaluate, so the destructured bindings end up captured in this
// module's local scope only.
//
// `setStatus(text | null)` drives the TUI footer busy indicator:
// a string shows the text with a spinner, `null` hides it.
//
// `setTitle(text | null)` sets or clears the session title, which the
// host persists across restarts. `getTitle()` returns the current title
// (or `null`): the persisted value at boot, updated locally on every
// `setTitle` — the workflow is the only writer, so no host round-trip.

export const { exit, setStatus } = globalThis.__frances_v1_stash__;

const { _setTitle, _initialTitle } = globalThis.__frances_v1_stash__;

let currentTitle = _initialTitle ?? null;

export function setTitle(text) {
  currentTitle = text ?? null;
  _setTitle(currentTitle);
}

export function getTitle() {
  return currentTitle;
}
